odoo/http.py

```diff
@@ -348,6 +360,116 @@ def route(route=None, **routing):
         by default for ``'json'``-type requests. See
         :ref:`CSRF Protection <csrf>` for more.
     """
+    def decorator(endpoint):
+        fname = f"<function {endpoint.__module__}.{endpoint.__name__}>"
+
+        # Sanitize the routing
+        assert 'type' not in routing or routing['type'] in ('http', 'json')
+        if route:
+            routing['routes'] = route if isinstance(route, list) else [route]
+        wrong = routing.pop('method', None)
+        if wrong is not None:
+            _logger.warning("%s defined with invalid routing parameter 'method', assuming 'methods'", fname)
+            routing['methods'] = wrong
+
+        @functools.wraps(endpoint)
+        def route_wrapper(self, *args, **params):
+            params_ok = filter_kwargs(endpoint, params)
+            params_ko = set(params) - set(params_ok)
+            if params_ko:
+                _logger.warning("%s called ignoring args %s", fname, params_ko)
+
+            result = endpoint(self, *args, **params_ok)
+            return result
+
+        route_wrapper.original_routing = routing
+        route_wrapper.original_endpoint = endpoint
+        return route_wrapper
+    return decorator
+
+def _generate_routing_rules(modules, nodb_only, converters=None):
+    def is_valid(cls):
+        path = cls.__module__.split('.')
+        return path[:2] == ['odoo', 'addons'] and path[2] in modules
+
+    def get_leaf_classes(cls):
+        result = []
+        for subcls in cls.__subclasses__():
+            if is_valid(subcls):
+                result.extend(get_leaf_classes(subcls))
+        if not result and is_valid(cls):
+            result.append(cls)
+        return result
+
+    def build_controllers():
+        highest_controllers = []
+        for module in modules:
+            highest_controllers.extend(Controller.children_classes.get(module, []))
+
+        for top_ctrl in highest_controllers:
+            leaf_controllers = list(unique(get_leaf_classes(top_ctrl)))
+            name = '{} (extended by {})'.format(
+                top_ctrl.__name__,
+                ', '.join(bot_ctrl.__name__ for bot_ctrl in leaf_controllers),
+            )
+            Ctrl = type(name, tuple(reversed(leaf_controllers)), {})
+            yield Ctrl()
+
+    for ctrl in build_controllers():
+        for method_name, method in inspect.getmembers(ctrl, inspect.ismethod):
+
+            # Skip this method if it is not @route decorated anywhere in
+            # the hierarchy
+            def is_method_a_route(cls):
+                return resolve_attr(cls, f'{method_name}.original_routing', None) is not None
+            if not any(map(is_method_a_route, type(ctrl).mro())):
+                continue
```

Comment:
Is it intended that the warning of [line 456](https://github.com/odoo/odoo/pull/78857/commits/530c5609c39a312930ef0fb114c05fc3e3ba49b2#diff-5f1b777eb0bc30f3f494ffe3e1522524cd0822f6c94c8644070a6db7c40b9c34R456) does not occur in this case?

Reply:
Yes it is expected. This case cover controller methods that are not routes, regular method such as foo in the example below:

    class MyController(http.Controller):
        @classmethod
        def foo(cls):
            return 5
            
        @route('/some/url', type='http')
        def some_url(self):
            return str(self.foo())